In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import r2_score,mean_squared_error
from sklearn.preprocessing import MinMaxScaler

In [2]:
datastartups=pd.read_csv("50_Startups.csv")
datastartups.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     object 
 4   Profit           50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


In [3]:
datastartups.head()
datastartups.describe()

R&D Spend  Administration  Marketing Spend         Profit
count      50.000000       50.000000        50.000000      50.000000
mean    73721.615600   121344.639600    211025.097800  112012.639200
std     45902.256482    28017.802755    122290.310726   40306.180338
min         0.000000    51283.140000         0.000000   14681.400000
25%     39936.370000   103730.875000    129300.132500   90138.902500
50%     73051.080000   122699.795000    212716.240000  107978.190000
75%    101602.800000   144842.180000    299469.085000  139765.977500
max    165349.200000   182645.560000    471784.100000  192261.830000

# Data preprocessing

In [4]:

# Renaming the column names with special charachters and space.
datastartups=datastartups.rename(columns={"R&D Spend":"RnDspent","Marketing Spend":"MarketingSpent"})

# using label.encoding for State 
stateencoder=LabelEncoder()
datastartups["State"]=stateencoder.fit_transform(datastartups["State"])
datastartups.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   RnDspent        50 non-null     float64
 1   Administration  50 non-null     float64
 2   MarketingSpent  50 non-null     float64
 3   State           50 non-null     int32  
 4   Profit          50 non-null     float64
dtypes: float64(4), int32(1)
memory usage: 1.9 KB


# Model Transformation - Scaling

In [5]:
scaler=MinMaxScaler()
datastartups[["RnDspent","Administration","MarketingSpent","State","Profit"]]=scaler.fit_transform(datastartups)
datastartups.describe()

RnDspent  Administration  MarketingSpent      State     Profit
count  50.000000       50.000000       50.000000  50.000000  50.000000
mean    0.445854        0.533345        0.447292   0.500000   0.548097
std     0.277608        0.213286        0.259208   0.416497   0.226974
min     0.000000        0.000000        0.000000   0.000000   0.000000
25%     0.241527        0.399260        0.274066   0.000000   0.424920
50%     0.441799        0.543661        0.450876   0.500000   0.525378
75%     0.614474        0.712221        0.634759   1.000000   0.704383
max     1.000000        1.000000        1.000000   1.000000   1.000000

In [6]:

# initialisisng x and y
y=datastartups.iloc[:,4]
x=datastartups.iloc[:,:-1]

# splitting dataset into test and train

xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.3,random_state=0)

# object creation and train the model
startModel=LinearRegression()
startModel.fit(xtrain,ytrain)

# Predicting the label

ytrainPredict=startModel.predict(xtrain)
ytestPredict=startModel.predict(xtest)

# Evaluating the output 

r2scoretrain=r2_score(ytrain,ytrainPredict)
r2scoretest=r2_score(ytest,ytestPredict)

print("Coefficient of determination r2Score-train =",r2scoretrain)
print("Coefficient of determination r2Score-test =",r2scoretest)

meansquarederrortrain=mean_squared_error(ytrain,ytrainPredict)
meansquarederrortest=mean_squared_error(ytest,ytestPredict)

print("Mean Squared Error for train =",meansquarederrortrain)
print("Mean Squared Error for test =",meansquarederrortest)

Coefficient of determination r2Score-train = 0.9515496105627431
Coefficient of determination r2Score-test = 0.9355139722149945
Mean Squared Error for train = 0.0028425081883885405
Mean Squared Error for test = 0.0019738493470693383


# Model performance Before Scaling the dataset

Coefficient of determination r2Score-train = 0.9515496105627431
Coefficient of determination r2Score-test = 0.9355139722149947
Mean Squared Error for train = 89637953.1399842
Mean Squared Error for test = 62244962.389464445

Before scaling the Coefficient of determination was good both during testing and training since its value is almost 95% that means this model is able to explain 95% of the variance in the data.
But there is a problem with the loss function - the loss function for a good perfoming model should be nearly zero but here it is high.

# Model performance  After transformation-scaling the Dataset

Coefficient of determination r2Score-train = 0.9515496105627431
Coefficient of determination r2Score-test = 0.9355139722149945
Mean Squared Error for train = 0.0028425081883885405
Mean Squared Error for test = 0.0019738493470693383

After scaling the Coefficient of determination remains good since its value is almost 95% that means this model is able to explain 95% of the variance in the data.
The loss function - the loss function for a good perfoming model should be nearly zero we can see after transforming the data it has reduced the MSE = loss function value to 0.002 train and 0.001 for test which is quite good.

So we can say our model after scaling is performing very well both in test and train and also expalining the variance with a very low residual error value which is our goal while creating a good model.
